properties.md

```diff
@@ -0,0 +1,140 @@
+},
+{
+    "type":  "olm.gvk",
+    "value": `{"group":"etcd.database.coreos.com","kind":"EtcdBackup","version":"v1beta2"}`,
+},
+{
+    "type":  "olm.label",
+    "value": `{"label":"testlabel"}`,
+}
+```
+
+Properties intentionally closely align with the format for [`dependencies.yaml`](https://github.com/operator-framework/operator-registry/blob/master/docs/design/operator-bundle.md#bundle-dependencies).
+
+Today, all properties are derived from other fields in a bundle. There is no way to write a new property `type` for a consumer (say, a newer version of the resolver, or an external tool that processes sets of bundles).
+
+### Other consumers
+
+Catalog and Bundles are consumed in many other contexts than the resolver. The package-server component presents catalog content for the purpose of selecting packages and channels, for example. 
+
+### Goals
+
+- Define the format for explicitly adding properties to a bundle.
+
+## Proposal
+
+A new file is accepted in the `metadata` folder for a bundle: `properties.(yaml|json)`
+
+Here is an example of `properties.yaml`:
+
+```yaml=
+properties:
+- type: olm.package
+  value:
+    packageName: prometheus
+    version: "0.27.0"
+- type: olm.gvk
+  value:
+    group: etcd.database.coreos.com
+    kind: EtcdCluster
+    version: v1beta2
+- type: olm.label
+  value: "LTS"
+```
+
+If properties are defined, they take precendence over the values derived from the bundle contents. In the above example, the `version` of the bundle will be considered `0.27.0` regardless of what the `ClusterServiceVersion`'s `spec.version` field indicates.
+
+### Validate properties
+
+Property validation is the domain of the consumer, not the producer. 
+
+However, `opm` will ship with validation for all properties that the OLM resolver understands, and they will be validated as part of the `opm alpha bundle validate` command and the `operator-sdk` bundle validation. 
+
+A future enhancement will discuss a formalized way to discover / publish validation rules for a given property `type`.
+
+### Generate default properties
+
+`opm` and `operator-sdk` will be capable of scaffolding the set of inferred properties from a given bundle. `opm` today infers a set of properties based on bundle content and stores them in the registry.
+
+### Initial Resolver-Consumer Properties
+
+This is a list of properties that the resolver consumer will read and understand.
+
+- `olm.package`: The package name and bundle version. Used as a target of version range dependencies.
+- `olm.gvk`: The GVK of a provided api. Used as the target of a required API dependency.
+- `olm.replaces`: The name of another bundle that this one replaces
+- `olm.skips`: A list of bundles that this one should skip
+- `olm.skipRange`: A semver range specifier; any bundle that falls in this range can skip to the current one.
+- `olm.label`: A string identifier for partitition bundles
+- `olm.deprecated`: A bundle with this property will never be picked by the OLM resolver
```

Comment:
Let's see if I understood this option. 
A user creates a project and builds a bundle for it such as 1.0.0. 

Then, these users make a PR with the bundle against https://github.com/operator-framework/community-operators/tree/master/community-operators and another one https://github.com/operator-framework/community-operators/tree/master/upstream-community-operators to publish it in OCP and operator.hub. 

After that, will the user able to push pr's against the bundles already published/merged and overwrite their config with its file? How it will works? Could you please describe here an example with the required steps/actions over how to deprecated the project to clarifies?

Reply:
Good question, and this is where the declarative config comes into play. 

You have a bundle image with a properties file like:

```yaml
properties:
 - type: olm.label
    value: LTS
 - type: olm.packageVersion
   value: 
    package: etcd
    version: 0.9.6 
```

When you add that to an index in the new format, those properties get persisted:

```json
{
  "schema": "olm.bundle",
  "bundleImage": "quay.io/etcd/bundle:@sha256:123",
  "properties": [
     {
      "type": "olm.label",
      "value": "LTS"
     },
    { 
      "type": "olm.packageVersion",
      "value": {
        "package": "etcd",
        "version": "0.9.6" 
      }
    }
  ]
}
```

If you later want to deprecate a bundle, you can go back to the _index_ and modify those properties there - without touching the bundle itself.

```json
{
  "schema": "olm.bundle",
  "bundleImage": "quay.io/etcd/bundle:@sha256:123",
  "properties": [
     { 
       "type": "olm.deprecated",
       "value": true,
     },
     {
      "type": "olm.label",
      "value": "LTS"
     },
    { 
      "type": "olm.packageVersion",
      "value": {
        "package": "etcd",
        "version": "0.9.6" 
      }
    }
  ]
}
```

The properties in the bundle are more or less "default properties" for when you add a bundle to an index. Once it's in an index, the index information is the source of truth.